File: srcs/pewpew.c
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   pewpew.c                                           :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: pitriche <pitriche@student.42.fr>          +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2019/09/22 15:55:59 by pitriche          #+#    #+#             */
/*   Updated: 2020/01/24 12:24:24 by pitriche         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include "doom_nukem.h"

void		pewpew(t_al *al)
{
	unsigned	i;
	t_rc_ray	ray;

	ft_bzero(&ray, sizeof(t_rc_hit));
	cast_ray(al, al->play.dir, &ray);
	i = 0;
	while (i < ray.nb_hits)
	{
		if (ray.hits[i].is_entity)
			if (ray.hits[i].ent.alive)
			{
				al->ent[ray.hits[i].ent.index].alive = 0;
				if (al->ent[ray.hits[i].ent.index].alive)
				{
					al->ent[ray.hits[i].ent.index].velx = 20 * al->sin[al->play.dir];
					al->ent[ray.hits[i].ent.index].vely = 20 * al->cos[al->play.dir];
					al->ent[ray.hits[i].ent.index].velz = 2;
				}
				return ;
			}
		i++;
	}
}

File: web version/src/pewpew.js
function piouPiou(al, i, ray) {
    const entity = al.ent[ray.hits[i].ent.index];

    if (entity.alive < 0) {
        entity.alive = 0;
        entity.anim = 0;
    }

    if (entity.alive) {
        entity.velx = 20 * al.sin[al.play.dir];
        entity.vely = 20 * al.cos[al.play.dir];
        if (entity.posz === al.sec[entity.csec].flHei) {
            entity.velz = 2;
        }
    }
}

function pewpew(al) {
    al.fireAnim = 0;

    const ray = { hits: [], nbHits: 0 };
    castRay(al, al.play.dir, ray);

    for (let i = 0; i < ray.nbHits; i++) {
        if (ray.hits[i].isEntity && ray.hits[i].ent.alive) {
            const entity = al.ent[ray.hits[i].ent.index];
            entity.alive -= al.play.dmg;
            piouPiou(al, i, ray);
            return;
        }
    }
}